Arduino/MiP_ProMini_Pack_Library/src/mip.cpp

```diff
@@ -1937,6 +1951,251 @@ int8_t MiP::rawGetUserData(uint8_t address, uint8_t& userData)
 
 
 
+void MiP::enableMiPDetectionMode(uint8_t id, uint8_t txPower)
+{
+    m_irId = id;
+
+    rawSetMiPDetectionMode(id, txPower);
+}
+
+void MiP::disableMiPDetectionMode()
+{
+    m_irId = MIP_IR_DETECTION_MODE_DISABLE;
+
+    // According to WowWee documentation, TX power must be between 1 and 120 even when disabling.
+    rawSetMiPDetectionMode(MIP_IR_DETECTION_MODE_DISABLE, 0x01);
+}
+
+bool MiP::isMiPDetectionModeEnabled()
+{
+    return m_irId > MIP_IR_DETECTION_MODE_DISABLE ? true : false;
+}
+
+int8_t MiP::readDetectedMiP(uint8_t& detectedMiP)
+{
+    // Fetch bytes from the Serial receive buffer and process any event data found within.
+    processAllResponseData();
+
+    if(!m_detectedMiP.detected)
+    {
+        return 0;
+    }
+
+    detectedMiP = m_detectedMiP.id;
+
+    // Now that it's been read, invalidate the last detected MiP.
+    m_detectedMiP.clear();
+
+    return 1;
+}
+
+// This internal protected method sends the set detection mode command with minimal error
+// handling. The error recovery happens at a higher level of the driver.
+void MiP::rawSetMiPDetectionMode(uint8_t id, uint8_t txPower)
+{
+    uint8_t command[1+2];
+
+    MIP_ASSERT( MIP_IR_DETECTION_MODE_DISABLE <= id && id <= 0xFF );
+    MIP_ASSERT( 0x01 <= txPower && txPower <= 0x78 );
+
+    command[0] = MIP_CMD_SET_DETECTION_MODE;
+    command[1] = id;
+    command[2] = txPower;
+
+    rawSend(command, sizeof(command));
+}
+
+void MiP::enableIRRemoteControl()
+{
+    verifiedIRRemoteControl(MIP_IR_REMOTE_CONTROL_ENABLE);
+}
+
+void MiP::disableIRRemoteControl()
+{
+    verifiedIRRemoteControl(MIP_IR_REMOTE_CONTROL_DISABLE);
+}
+
+bool MiP::isIRRemoteControlEnabled()
+{
+    const uint8_t remoteControlEnabled[1] = { MIP_CMD_GET_IR_REMOTE_CONTROL };
+    uint8_t       response[1+1];
+    size_t        responseLength;
+    int8_t        result;
+
+    result = rawReceive(remoteControlEnabled, sizeof(remoteControlEnabled), response, sizeof(response), responseLength);
+
+    if (result)
+    {
+        return result;
+    }
+    if (responseLength != sizeof(response) ||
+        response[0] != MIP_CMD_GET_IR_REMOTE_CONTROL)
+    {
+        return MIP_ERROR_BAD_RESPONSE;
+    }
+
+    return response[1] == MIP_IR_REMOTE_CONTROL_ENABLE ? true : false;
+}
+
+void MiP::sendIRDongleCode(uint8_t sendCode[], uint8_t dataNumbers, uint8_t transmitPower)
+{
+    // According to WowWee documentation, indicate number of valid bits in transmission.
+    MIP_ASSERT( dataNumbers == 0x02 || dataNumbers == 0x03 || dataNumbers == 0x04 );
+
+    // Check for valid transmitPower values.  Must be between 1 and 120.
+    MIP_ASSERT( 0x01 <= transmitPower && transmitPower <= 0x78 );
+
+    // Check for valid codes in a three-byte send.
+    if(dataNumbers == 0x03 && sendCode[3] != 0xFF)
```

Comment:
Suggestion: I think this can be rewritten as a single assert.

```
MIP_ASSERT ( dataNumber != 3 || sendCode[3] == 0xFF || (0x00 <= sendCode[1] && sendCode[1] <= 0x03) );
```

Reply:
OK, then how about this for a four-byte send?

```
    MIP_ASSERT( dataNumbers != 4 || (sendCode[2] == 0xFF && sendCode[3] == 0xFF) || ((0x00 <= sendCode[0] && sendCode[0] <= 0x03) && (0x00 <= sendCode[1] && sendCode[1] <= 0x03));
```

Reply:
I think that is right.